### ISSUE TITLE
cb_explore returns always first arm when predicting when no context variables are passed

### ISSUE THREAD

**Issue by vmolina** [JohnLangford/vowpal_wabbit#1436]
Hello, 
I am trying to use the contextual bandit build on Vowpal Wabbit. 
I train the model with the following code
`vw -d train_temp.data -f tmp_fake.model --cb_explore 4 --epsilon 0.4 --cb_type dm -l 0.001 --power_t 0 --invert_hash tmp_fake.txt`
This is an example of the training file. Arm 3 has the lower cost (stable 0.1) i am passing no context.
```
1:1.0:0.25 |
4:1.0:0.25 |
4:1.0:0.25 |
3:0.1:0.25 |
3:0.1:0.25 |
3:0.1:0.25 |
3:0.1:0.25 |
2:1.0:0.25 |
4:1.0:0.25 |
1:1.0:0.25 |
```
When I look at the weights it has the following result.
```

Version 8.5.0
Id 
Min label:0
Max label:1
bits:18
lda:0
0 ngram:
0 skip:
options: --hash_seed 0 --cb_explore 4 --epsilon 0.4 --psi 1 --cb 4 --cb_type dm --csoaa 4 --link identity
Checksum: 4099937526
:0
Constant:202096:0.999999
Constant[1]:202097:0.999999
Constant[2]:202098:0.0999995
Constant[3]:202099:0.999999
```

The model has clearly learn that the third arm has the lower cost, but when I predict I always obtain arm 1.
The testing file is 
```
|
|
|
|
|
|
|
|
|
|
```
And the output of probabilities is
`vw -d test.txt -i tmp_fake.model -p /dev/stdout`

```
0.700000 0.100000 0.100000 0.100000 
0.700000 0.100000 0.100000 0.100000 
0.700000 0.100000 0.100000 0.100000 
0.700000 0.100000 0.100000 0.100000 
0.700000 0.100000 0.100000 0.100000 
0.700000 0.100000 0.100000 0.100000 
0.700000 0.100000 0.100000 0.100000 
0.700000 0.100000 0.100000 0.100000 
0.700000 0.100000 0.100000 0.100000 
0.700000 0.100000 0.100000 0.100000 

```

**Comment by JohnLangford** [comment]
I can reproduce.   The problem seems to be specific to DM, and it appears there is a failure to actually evaluate during testing since --audit produces nothing extra.   The problem seems to start here:
https://github.com/JohnLangford/vowpal_wabbit/blob/master/vowpalwabbit/gen_cs_example.h#L59

**Comment by matanster** [comment]
@vmolina would it be fair to ask here what would the probabilities (as in the input file) very semantically mean?

>1:1.0:0.25 |
4:1.0:0.25 |
4:1.0:0.25 |
3:0.1:0.25 |
3:0.1:0.25 |
3:0.1:0.25 |
3:0.1:0.25 |
2:1.0:0.25 |
4:1.0:0.25 |
1:1.0:0.25 |

Does each line simply function as an aggregate of observations collected from history over consecutive periods of time, which the CB model would simply use for learning one by one in the batch training mode that this all is part of? or is the interpretation very different.

What does it mean that `1:1.0` was observed with probability `0.25`, and later observed with the same probability after 9 lines? Where does the remaining 0.75 probability go? What does it mean that `3:0.1` was observed 4 times in succession? 

we're grappling here with mapping this very strictly to a real world scenario...

**Comment by matanster** [comment]
@vmolina thanks for commenting. 

> So each line is a single observation.

Where does a probability (here 0.25) come from given that each line is a single deterministic observation?

**Comment by vmolina** [comment]
From the wiki page

> Here each line is a separate example and each takes the form:
> 
> `action:cost:probability | features`
> Where 
>  - action is the id of the action taken where we observed the cost (a positive integer in {1, k})
>  - cost is the cost observed for this action (floating point, lower is better)
>  - probability is the probability (floating point, in [0..1]) of the exploration policy to choose this action when collecting the data
>  - features are the list of all features for this example specified as usual for classification/regression problem with vw


As the actions are explored uniformly  and there are 4 action we set it to 0.25.

**Comment by matanster** [comment]
In our scenario we're dealing with Vowpal with online simulations not with real data (which we don't have yet). And we don't really get this part ― If you're training over the entire data epoch that comprises your training data, and the entire data epoch comprising your history is uniformly distributed exploration, why would you have duplicate lines in your data? 

I appreciate this is outside the scope of your use cases, so thanks for your commenting if you do.

**Comment by JohnLangford** [comment]
This is fixed in #1440 .  The core issue was here: https://github.com/JohnLangford/vowpal_wabbit/pull/1440/files#diff-4762f2ebc108c24b1f4f07229f04a342R59 where the code inappropriately diverted to the 'else' clause.  That's because we regarded DM as such a baseline that we didn't test this....   I'll add a test.
